sdk/src/binary.rs

```diff
@@ -0,0 +1,283 @@
+        let v32: u32 = v.into();
+        self.0 = self
+            .env()
+            .binary_put(self.0.to_tagged(), i.into(), v32.into())
+            .in_env(self.env());
+    }
+
+    #[inline(always)]
+    pub fn get(&self, i: u32) -> u8 {
+        let res32: u32 = self
+            .env()
+            .binary_get(self.0.to_tagged(), i.into())
+            .try_into()
+            .unwrap();
+        res32.try_into().unwrap()
+    }
+
+    #[inline(always)]
+    pub fn del(&mut self, i: u32) {
+        self.0 = self
+            .env()
+            .binary_del(self.0.to_tagged(), i.into())
+            .in_env(self.env());
+    }
+
+    #[inline(always)]
+    pub fn len(&self) -> u32 {
+        self.env()
+            .binary_len(self.0.to_tagged())
+            .try_into()
+            .unwrap()
+    }
+
+    pub fn push(&mut self, x: u8) {
+        let x32: u32 = x.into();
+        self.0 = self
+            .env()
+            .binary_push(self.0.to_tagged(), x32.into())
+            .in_env(self.env());
+    }
+
+    pub fn pop(&mut self) {
+        self.0 = self.env().binary_pop(self.0.to_tagged()).in_env(self.env());
+    }
+
+    pub fn front(&self) -> u8 {
+        let res32: u32 = self
+            .env()
+            .binary_front(self.0.to_tagged())
+            .try_into()
+            .unwrap();
+        res32.try_into().unwrap()
+    }
+
+    pub fn back(&self) -> u8 {
+        let res32: u32 = self
+            .env()
+            .binary_back(self.0.to_tagged())
+            .try_into()
+            .unwrap();
+        res32.try_into().unwrap()
+    }
+
+    pub fn insert(&mut self, i: u32, x: u8) {
+        let x32: u32 = x.into();
+        self.0 = self
+            .env()
+            .binary_insert(self.0.to_tagged(), i.into(), x32.into())
+            .in_env(self.env());
+    }
+
+    pub fn append(&mut self, other: &Binary) {
+        self.0 = self
+            .env()
+            .binary_append(self.0.to_tagged(), other.0.to_tagged())
+            .in_env(self.env());
+    }
+}
+
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct FixedLengthBinary<const N: u32>(Binary);
+
+impl<const N: u32> TryFrom<EnvVal> for FixedLengthBinary<N> {
+    type Error = ();
+
+    #[inline(always)]
+    fn try_from(ev: EnvVal) -> Result<Self, Self::Error> {
+        let obj: EnvObj = ev.try_into()?;
+        obj.try_into()
+    }
+}
+
+impl<const N: u32> TryFrom<EnvObj> for FixedLengthBinary<N> {
+    type Error = ();
+
+    #[inline(always)]
+    fn try_from(obj: EnvObj) -> Result<Self, Self::Error> {
+        let bin: Binary = obj.try_into()?;
+        bin.try_into()
+    }
+}
+
+impl<const N: u32> TryFrom<Binary> for FixedLengthBinary<N> {
+    type Error = ();
+
+    #[inline(always)]
+    fn try_from(bin: Binary) -> Result<Self, Self::Error> {
+        if bin.len() == N {
+            Ok(Self(bin))
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl<const N: u32> AsRef<Binary> for FixedLengthBinary<N> {
```

Comment:
I think it'd also be worth impl `Deref<Binary>`, although importantly not `DerefMut`. So that all the read-only fns of Binary are available on `FixedLengthBinary`.

Reply:
This is an [anti-pattern](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html). I introduced new traits `FixedLengthBinary` and `VariableLengthBinary` which include the functions relevant to those to categories. `Binary` implements `FixedLengthBinary` and `VariableLengthBinary`. `ArrayBinary<N>` implements `FixedLengthBinary` as a pass-through to `Binary`, except that it returns the static length `N` instead of passing-through for `len`.

Now you can operate on either of them generically by using the appropriate trait bounds.